Add FileSystemOptions and use in FileResolver

MarkArbogast opened the pull request:
Addresses: #2550 

Signed-off-by: Mark Arbogast <mark.t.arbogast@gmail.com>

vietj commented on src/main/java/io/vertx/core/VertxOptions.java:
```diff
@@ -705,7 +717,7 @@ public VertxOptions setAddressResolverOptions(AddressResolverOptions addressReso
    */
   public boolean isFileResolverCachingEnabled() {
-    return fileResolverCachingEnabled;
+    return fileSystemOptions.isFileResolverCachingEnabled();
```
missing null check

vietj commented on src/main/java/io/vertx/core/VertxOptions.java:
```diff
@@ -715,7 +727,7 @@ public boolean isFileResolverCachingEnabled() {
    */
   public VertxOptions setFileResolverCachingEnabled(boolean fileResolverCachingEnabled) {
-    this.fileResolverCachingEnabled = fileResolverCachingEnabled;
+    this.fileSystemOptions.setFileResolverCachingEnabled(fileResolverCachingEnabled);
```
missing null check

vietj commented:
looks good, it should care of null file system options on the vertx options

vietj commented:
we also want to deprecate the existing options on VertxOptions and say to use the nested file system options.

MarkArbogast commented:
Added null checks and a unit test to verify them. Also deprecated `isFileResolverCachingEnabled` and `setFileResolverCachingEnabled`.

pmlopes commented:
Question: should we also drop the system properties too? Does it make sense to allow changes at runtime? What takes precedence options or system properties?

vietj commented:
@pmlopes that would be a breaking change, system properties should configure the default property values

vietj commented on src/main/java/io/vertx/core/file/FileSystemOptions.java:
```diff
@@ -0,0 +1,117 @@
+
+/*
+ * Copyright (c) 2011-2017 Contributors to the Eclipse Foundation
```
2018

vietj commented on src/main/java/io/vertx/core/file/impl/FileResolver.java:
```diff
@@ -50,34 +51,38 @@
   private File cacheDir;
   private Thread shutdownHook;
-  private final boolean enableCaching;
+  private final FileSystemOptions fileSystemOptions;
```
keep the `enableCaching` field and instead assign its value from the provided options

vietj commented on src/main/java/io/vertx/core/file/impl/FileResolver.java:
```diff
@@ -104,7 +109,7 @@ public File resolveFile(String fileName) {
       file = new File(cwd, fileName);
     }
-    if (!ENABLE_CP_RESOLVING) {
+    if (!fileSystemOptions.isClassPathResolvingEnabled()) {
```
add a final `enableCpResolving` field to the class instead

vietj commented:
I've added a couple of extra comments, I think it needs to be also documented

MarkArbogast commented:
Switched to use boolean variables instead of options in `FileResolver`. Updated documentation. Updated 2017 to 2018 in `FileSystemOptions` copyright comment.

MarkArbogast commented:
Updated 2017 to 2018 in `FileSystemOptions` copyright comment.

MarkArbogast commented:
Hi @vietj, anything else to fix for this PR?

vietj commented on src/test/java/io/vertx/test/core/VertxOptionsTest.java:
```diff
@@ -416,4 +424,15 @@ public void testJsonOptions() {
+      options.isFileResolverCachingEnabled();
+      options.setFileResolverCachingEnabled(true);
+    } catch (NullPointerException e) {
+      fail("Should handle null FileSystemOptions");
```
just let it go through and let the NPE fail the test

MarkArbogast commented:
Updated `testNullFileSystemOptions` to fail directly instead of catching NPE.

vietj commented:
thanks for the contrib!
